pg_search/src/postgres/storage/fsm.rs

```diff
@@ -34,413 +35,1263 @@ enum FSMBlockKind {
+/// the largest existing key (transaction id) an use that slot.
+///
+/// If that key is `>=` to the new transaction id we're trying to insert, then we simply use that
+/// slot's information.
+///
+/// If the transaction id we're trying to insert is larger, we **directly modify the existing key**
+/// to be the new transaction id.  The tree will maintain its balance as the new key is still greater
+/// than its predecessors.
+///
+/// Then extension happens as normal.
+///
+/// This has a side effect of moving either existing blocks or the set of new blocks to the future.
+/// For existing blocks this means they potentially won't be reused as soon as they could.
+///
+/// In practice, it'll be quite challenging to fill the tree -- perhaps even impossible.  It would
+/// require hundreds of concurrent merges to happen without any new segments being created.  This is
+/// not a scenario that can actually happen.  Merging is what returns blocks to the FSM and merging
+/// only happens against segments.
+///
+/// # Drain
+///
+/// Draining is the process of asking the FSM for free blocks.  The caller asks for `n` blocks and
+/// the FSM returns an iterator of blocks that can be reused by the current transaction.  "Current
+/// transaction" here literally means the result of `pg_sys::GetCurrentFullTransactionId()`, but
+/// _could_ be a different, father-in-the-past, transaction.
+///
+/// The drain process takes a shared lock on the root page and finds the entry that is less-than-or-equal-to
+/// the current transaction id.  A transaction can only reuse blocks related to the same or older
+/// transactions.
+///
+/// If no such key exists, [`FreeSpaceManager::drain`] returns an empty iterator.
+///
+/// Otherwise, the blocklist associated with that entry's slot is consumed, up to `n` blocks.  The
+/// locking through the blocklist here is conditional.  If an exclusive lock can't be acquired, the
+/// algorithm restarts, now looking for the next smallest transaction id.  This continues until `n`
+/// blocks have been found or the tree no longer has any keys that satisfy the condition.
+///
+/// This process of restarting with the next smallest transaction id also happens if the xid being
+/// evaluated doesn't contain enough blocks to satisfy `n`.
+///
+/// When draining, if an individual block in the blocklist (that isn't the first block) becomes empty,
+/// it is unlinked from the blocklist and returned to the FSM to be used in a future transaction.
+/// This helps to keep the FSM as small as possible throughout its life.
+///
+/// If, while draining, an entire blocklist is consumed for a transaction id, then that transaction
+/// id is removed from the tree.  This requires an exclusive lock on the tree itself and must happen
+/// outside holding any other locks.  It would be possible for two concurrent transactions to try
+/// and delete the same xid entry.  This is fine -- one will win and the other will happily think it won.
```

Comment:
Would it be worth doing this during `extend` instead, to speed up `drain`?

Reply:
> `extend` ... could in theory scan everything and clean these up if needed.

`extend` is still called quite frequently (tho I think we've got it batched as well as we can now) and again, `extend` only knows about its xid slot.  Teaching it to (sometimes?) do a scan of the whole FSM structure just seems like extra complication.  I'm not a fan of this idea at all.